File: src/original-c#-camera/RealTimeProcessingCamera.cs
using AVFoundation;
using CoreFoundation;
using CoreMedia;
using CoreVideo;
using Foundation;
using System;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace RivaDigitalApp.iOS
{

    public class RealTimeProcessingCamera : ICamera
    {
        private TaskCompletionSource<object> recordingTask;
        private CancellationToken externalCancellationToken;
        private AVCaptureSession session;
        private RealTimePPGSignalsProcessor ppgSignalsProcessor;
        private DispatchQueue dispatchQueue;

        public bool IsCameraSupported()
        {
            return true;
        }

        public bool IsFullHardwareSupported()
        {
            return true;
        }

        public IPPGSignalsCollector PpgSignalsCollector { get; private set; }

        public Task StartRecordingAsync(CancellationToken cancellationToken)
        {
            this.externalCancellationToken = cancellationToken;
            this.externalCancellationToken.Register(OnCancelRequested);

            EnsureRecordingTask();

            if (!SetupCaptureSession())
            {
                FinalizeRecordingTask(new Exception("Can not initialize recording"));
            }

            if (recordingTask != null)
            {
                return recordingTask.Task;
            }

            return Task.FromException(new Exception("Can not initialize recording"));
        }

        public CameraSettings CameraSettings { get; } = new CameraSettings(
            frameRate: 60,
            frameWidth: 1280,
            frameHeight: 720,
            duration: TimeSpan.FromSeconds(30));

        private void EnsureRecordingTask()
        {
            if (recordingTask != null)
            {
                FinalizeRecordingTask();
            }

            recordingTask = new TaskCompletionSource<object>();
        }

        private bool SetupCaptureSession()
        {
            session = new AVCaptureSession();

            AVCaptureDevice captureDevice = GetCaptureDevice();
            if (captureDevice == null)
            {
                Console.WriteLine("No captureDevice - this won't work on the simulator, try a physical device");
                return false;
            }

            var input = AVCaptureDeviceInput.FromDevice(captureDevice);
            if (input == null)
            {
                Console.WriteLine("No input - this won't work on the simulator, try a physical device");
                return false;
            }

            session.AddInput(input);
            captureDevice.LockForConfiguration(out var error);
            if (error != null)
            {
                Console.WriteLine(error);
                captureDevice.UnlockForConfiguration();
                return false;
            }

            //captureDevice.TorchMode = AVCaptureTorchMode.On;
            (var activeFormat, var frameRate, var frameWidth, var frameHeight) = GetCaptureDeviceFormat(captureDevice);

            captureDevice.ActiveFormat = activeFormat;
            var cmTime = new CMTime(10, frameRate * 10);

            captureDevice.ActiveVideoMinFrameDuration = cmTime;
            captureDevice.ActiveVideoMaxFrameDuration = cmTime;

            captureDevice.UnlockForConfiguration();

            var bufferAttr = new CVPixelBufferAttributes
            {
                PixelFormatType = CVPixelFormatType.CV420YpCbCr8BiPlanarFullRange,
            };

            using (var output = new AVCaptureVideoDataOutput { WeakVideoSettings = bufferAttr.Dictionary })
            {
                dispatchQueue = new DispatchQueue("myQueue");
                PpgSignalsCollector = PPGSignalsCollector.CreateiOsCollector(frameWidth, frameHeight);
                ppgSignalsProcessor = new RealTimePPGSignalsProcessor(PpgSignalsCollector);
                output.SetSampleBufferDelegateQueue(ppgSignalsProcessor, dispatchQueue);
                session.AddOutput(output);
            }

            session.StartRunning();

            session.BeginConfiguration();
            captureDevice.LockForConfiguration(out var nSError);
            if (nSError != null)
            {
                Console.WriteLine(nSError);
                captureDevice.UnlockForConfiguration();
                return false;
            }

            if (captureDevice.TorchAvailable)
            {
                captureDevice.TorchMode = AVCaptureTorchMode.On;
            }

            if (captureDevice.FlashAvailable)
            {
                captureDevice.FlashMode = AVCaptureFlashMode.On;
            }

            captureDevice.UnlockForConfiguration();
            session.CommitConfiguration();
            Task.Delay(CameraSettings.Duration).ContinueWith((t) =>
            {
                FinalizeRecordingTask();
            }, externalCancellationToken);

            return true;
        }

        private static AVCaptureDevice GetCaptureDevice()
        {
            //Telephoto camera is closer to flashlight, try to use it.
            AVCaptureDevice device = AVCaptureDevice.GetDefaultDevice(AVCaptureDeviceType.BuiltInTelephotoCamera, AVMediaType.Video, AVCaptureDevicePosition.Back);
            if (device == null)
            {
                //Otherwise try to use back camera.
                device = AVCaptureDevice.DevicesWithMediaType(AVMediaType.Video).FirstOrDefault(d => d.Position == AVCaptureDevicePosition.Back);
            }
            return device;
        }

        private (AVCaptureDeviceFormat Format, int FrameRate, int FrameWdith, int FrameHeight) GetCaptureDeviceFormat(AVCaptureDevice captureDevice)
        {
            var formats = captureDevice.Formats
                .Where(f => f.FormatDescription.MediaSubType == (uint)CVPixelFormatType.CV420YpCbCr8BiPlanarFullRange)
                .Select(PrepareInfo)
                .ToList();

            var matched = formats
                .Where(x =>
                    x.FrameRate >= CameraSettings.FrameRate &&
                    x.FrameWidth >= CameraSettings.FrameWidth &&
                    x.FrameWidth >= CameraSettings.FrameHeight)
                .OrderBy(x => x.FrameWidth * x.FrameHeight)
                .ThenBy(x => x.FrameRate)
                .ToList();
            return matched.First();

            (AVCaptureDeviceFormat Format, int FrameRate, int FrameWidth, int FrameHeight) PrepareInfo(AVCaptureDeviceFormat format)
            {
                var videoDimensions = ((CMVideoFormatDescription)format.FormatDescription).Dimensions;
                return (format, (int) format.VideoSupportedFrameRateRanges.First().MaxFrameRate, videoDimensions.Width, videoDimensions.Height);
            }
        }

        #region Release resources

        // FinalizeRecordingTask can be called in separate threads,
        // so we have to sync these threads to make it safe.
        private readonly SemaphoreSlim semaphoreFinalize = new SemaphoreSlim(1, 1);

        private async void FinalizeRecordingTask()
        {
            await semaphoreFinalize.WaitAsync();

            try
            {
                session?.StopRunning();

                if (ppgSignalsProcessor != null)
                {
                    await ppgSignalsProcessor.ProcessingToCompletionAsync();

                    if (ppgSignalsProcessor != null)
                    {
                        ppgSignalsProcessor.Dispose();
                        ppgSignalsProcessor = null;
                    }
                }

                if (recordingTask != null && !recordingTask.Task.IsCompleted)
                {
                    recordingTask.SetResult(true);
                }
                recordingTask = null;

                FreeResources();
            }
            finally
            {
                semaphoreFinalize.Release();
            }
        }

        private void FinalizeRecordingTask(Exception e)
        {
            semaphoreFinalize.Wait();

            try
            {
                session?.StopRunning();

                if (ppgSignalsProcessor != null)
                {
                    ppgSignalsProcessor.TerminateProcessing();
                    ppgSignalsProcessor.Dispose();
                    ppgSignalsProcessor = null;
                }

                if (recordingTask != null && !recordingTask.Task.IsCompleted)
                {
                    recordingTask.SetException(e);
                }
                recordingTask = null;

                FreeResources();
            }
            finally
            {
                semaphoreFinalize.Release();
            }
        }

        private void OnCancelRequested()
        {
            FinalizeRecordingTask(new Exception("Operation canceled"));
        }

        private void FreeResources()
        {
            if (session != null)
            {
                session.Dispose();
                session = null;
            }

            if (dispatchQueue != null)
            {
                dispatchQueue.Dispose();
                dispatchQueue = null;
            }
        }

        #endregion
    }
}

File: src/swift-camera/RealTimeProcessingCamera.swift
import Foundation
import AVFoundation
import CoreMedia
import CoreVideo

// Swift equivalent of RealTimeProcessingCamera.cs
class RealTimeProcessingCamera: ICamera {
    private var recordingTask: Task<Void, Error>?
    private var externalCancellationToken: CancellationToken?
    private var session: AVCaptureSession?
    private var ppgSignalsProcessor: RealTimePPGSignalsProcessor?
    private var dispatchQueue: DispatchQueue?
    
    // Semaphore for thread-safe finalization
    private let semaphoreFinalize = DispatchSemaphore(value: 1)
    
    var ppgSignalsCollector: IPPGSignalsCollector?
    
    let cameraSettings = CameraSettings(
        frameRate: 60,
        frameWidth: 1280,
        frameHeight: 720,
        duration: 30.0
    )
    
    func isCameraSupported() -> Bool {
        return true
    }
    
    func isFullHardwareSupported() -> Bool {
        return true
    }
    
    func startRecordingAsync(cancellationToken: CancellationToken) async throws {
        self.externalCancellationToken = cancellationToken
        cancellationToken.register { [weak self] in
            self?.onCancelRequested()
        }
        
        ensureRecordingTask()
        
        if !setupCaptureSession() {
            throw NSError(domain: "RealTimeProcessingCamera", code: 1, userInfo: [NSLocalizedDescriptionKey: "Cannot initialize recording"])
        }
        
        if let recordingTask = recordingTask {
            try await recordingTask.value
        } else {
            throw NSError(domain: "RealTimeProcessingCamera", code: 2, userInfo: [NSLocalizedDescriptionKey: "Cannot initialize recording"])
        }
    }
    
    private func ensureRecordingTask() {
        if recordingTask != nil {
            finalizeRecordingTask()
        }
        
        recordingTask = Task {
            // Task will be completed by the timer or cancellation
            try await withCheckedThrowingContinuation { continuation in
                // Set up timer to complete after duration
                DispatchQueue.main.asyncAfter(deadline: .now() + cameraSettings.duration) {
                    continuation.resume()
                }
            }
        }
    }
    
    private func setupCaptureSession() -> Bool {
        session = AVCaptureSession()
        
        guard let captureDevice = getCaptureDevice() else {
            print("No captureDevice - this won't work on the simulator, try a physical device")
            return false
        }
        
        guard let input = try? AVCaptureDeviceInput(device: captureDevice) else {
            print("No input - this won't work on the simulator, try a physical device")
            return false
        }
        
        session?.addInput(input)
        
        do {
            try captureDevice.lockForConfiguration()
        } catch {
            print("Error locking device for configuration: \(error)")
            return false
        }
        
        // Get capture device format
        guard let formatInfo = getCaptureDeviceFormat(captureDevice) else {
            captureDevice.unlockForConfiguration()
            return false
        }
        
        captureDevice.activeFormat = formatInfo.format
        let cmTime = CMTime(value: 10, timescale: CMTimeScale(formatInfo.frameRate * 10))
        captureDevice.activeVideoMinFrameDuration = cmTime
        captureDevice.activeVideoMaxFrameDuration = cmTime
        
        captureDevice.unlockForConfiguration()
        
        // Set up video output
        let output = AVCaptureVideoDataOutput()
        output.videoSettings = [
            kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_420YpCbCr8BiPlanarFullRange
        ]
        
        dispatchQueue = DispatchQueue(label: "myQueue")
        ppgSignalsCollector = PPGSignalsCollector.createiOsCollector(width: formatInfo.frameWidth, height: formatInfo.frameHeight)
        ppgSignalsProcessor = RealTimePPGSignalsProcessor(ppgSignalsCollector: ppgSignalsCollector!)
        
        output.setSampleBufferDelegate(ppgSignalsProcessor, queue: dispatchQueue)
        session?.addOutput(output)
        
        session?.startRunning()
        
        // Configure torch and flash
        session?.beginConfiguration()
        
        do {
            try captureDevice.lockForConfiguration()
        } catch {
            print("Error locking device for configuration: \(error)")
            captureDevice.unlockForConfiguration()
            return false
        }
        
        if captureDevice.isTorchAvailable {
            captureDevice.torchMode = .on
        }
        
        if captureDevice.isFlashAvailable {
            captureDevice.flashMode = .on
        }
        
        captureDevice.unlockForConfiguration()
        session?.commitConfiguration()
        
        // Set up completion timer
        DispatchQueue.main.asyncAfter(deadline: .now() + cameraSettings.duration) { [weak self] in
            self?.finalizeRecordingTask()
        }
        
        return true
    }
    
    private func getCaptureDevice() -> AVCaptureDevice? {
        // Try telephoto camera first (closer to flashlight)
        if let device = AVCaptureDevice.default(.builtInTelephotoCamera, for: .video, position: .back) {
            return device
        }
        
        // Otherwise try back camera
        return AVCaptureDevice.default(.builtInWideAngleCamera, for: .video, position: .back)
    }
    
    private func getCaptureDeviceFormat(_ captureDevice: AVCaptureDevice) -> (format: AVCaptureDevice.Format, frameRate: Int, frameWidth: Int, frameHeight: Int)? {
        let formats = captureDevice.formats.compactMap { format -> (format: AVCaptureDevice.Format, frameRate: Int, frameWidth: Int, frameHeight: Int)? in
            guard format.formatDescription.mediaSubType == kCVPixelFormatType_420YpCbCr8BiPlanarFullRange else {
                return nil
            }
            
            let dimensions = CMVideoFormatDescriptionGetDimensions(format.formatDescription)
            let maxFrameRate = Int(format.videoSupportedFrameRateRanges.first?.maxFrameRate ?? 0)
            
            return (format: format, frameRate: maxFrameRate, frameWidth: Int(dimensions.width), frameHeight: Int(dimensions.height))
        }
        
        let matched = formats.filter { info in
            info.frameRate >= cameraSettings.frameRate &&
            info.frameWidth >= cameraSettings.frameWidth &&
            info.frameHeight >= cameraSettings.frameHeight
        }.sorted { lhs, rhs in
            let lhsArea = lhs.frameWidth * lhs.frameHeight
            let rhsArea = rhs.frameWidth * rhs.frameHeight
            if lhsArea != rhsArea {
                return lhsArea < rhsArea
            }
            return lhs.frameRate < rhs.frameRate
        }
        
        return matched.first
    }
    
    // MARK: - Resource Management
    
    private func finalizeRecordingTask() {
        semaphoreFinalize.wait()
        defer { semaphoreFinalize.signal() }
        
        session?.stopRunning()
        
        if let processor = ppgSignalsProcessor {
            Task {
                await processor.processingToCompletionAsync()
                
                DispatchQueue.main.async { [weak self] in
                    self?.ppgSignalsProcessor = nil
                }
            }
        }
        
        recordingTask = nil
        freeResources()
    }
    
    private func finalizeRecordingTask(error: Error) {
        semaphoreFinalize.wait()
        defer { semaphoreFinalize.signal() }
        
        session?.stopRunning()
        
        ppgSignalsProcessor?.terminateProcessing()
        ppgSignalsProcessor = nil
        
        recordingTask?.cancel()
        recordingTask = nil
        
        freeResources()
    }
    
    private func onCancelRequested() {
        let error = NSError(domain: "RealTimeProcessingCamera", code: 3, userInfo: [NSLocalizedDescriptionKey: "Operation canceled"])
        finalizeRecordingTask(error: error)
    }
    
    private func freeResources() {
        session = nil
        dispatchQueue = nil
    }
}
